flax/linen/module.py

```diff
@@ -790,9 +791,52 @@ def make_rng(self, name: str) -> PRNGKey:
       raise ValueError("Can't use RNGs on unbound modules")
     return self.scope.make_rng(name)
 
+  def bind(self, variables: VariableDict, *args, rngs: RNGSequences = None,
+           mutable: CollectionFilter = False):
+    """Creates an interactive Module instance by binding variables and RNGs.
+
+    bind provides an "interactive" instance of a Module directly without
+    transforming a function with ``apply``. This is particulary useful for debugging
+    and interactive use cases like notebooks where a function would limit the ability
+    split up code into different cells.
+
+    Once the variables (and optionally RNGs) are bound to a ``Module`` it becomes a
+    stateful object. Note that idiomatic JAX is functional and therefore an interactive
+    instance does not mix well well with vanilla JAX APIs. Therefore, we recommend using
+    ``apply`` when code should be reusable and compatible across the JAX software ecosystem.
+
```

Comment:
It would really complement this docstring nicely, to have a very short HOWTO showing how to convert code that uses `bind` to code that uses `nn.apply` (by simply putting code inside a function).

Basically showing the autoencoder example that you have in the docstring and then on the RHS of the diff showing:
```
def f(model, x):
  z = model.encode(x)
  return model.decode(z)

ae = AutoEncoder()
x_reconstructed = ae.apply(f)(variables, x)
```

wdyt?

Reply:
Basically, then the docstring guides you out of your hole, once you want to use JAX transformations after you've tinkered in Colab for a while.

Reply:
Yes I agree, I would like to think a bit more about what we would put into an apply/bind howto so I'll make an issue for it